Turning the CSRF component into a middleware

burzum opened the pull request:
What the title says.

burzum committed:
Started on turning the CSRF component into a middleware

burzum committed:
Adding tests for the CSRF middleware

markstory commented on src/Http/Middleware/CsrfProtectionMiddleware.php:
```diff
@@ -0,0 +1,184 @@
+     */
+    public function __construct(array $config = [])
+    {
+        $this->_config = $this->_defaultConfig + $config;
```
This merge is backwards.

robertpustulka replied:
Can't you use `InstanceConfigTrait`?

burzum replied:
Why do we *need* this? Instead of using yet another config array I'm thinking about changing the config to setters.

robertpustulka replied:
@burzum the trait already has getters and setters for config :)

markstory replied:
I'm not sure we need the config trait. I don't see many scenarios where configuration needs to be modified after the middleware is created.

burzum replied:
I haven't added it for exactly the reason @markstory mentioned. I don't the a need to add it.

markstory commented on src/Http/Middleware/CsrfProtectionMiddleware.php:
```diff
@@ -0,0 +1,184 @@
+        $cookies = $request->getCookieParams();
+        $cookieData = null;
+        if (isset($cookies[$this->_config['cookieName']])) {
+            $cookieData = isset($cookies[$this->_config['cookieName']]);
```
Won't this make the cookieData always true?

chrisShick replied:
I think this was just a copy&paste error!

burzum replied:
Yes

markstory commented on src/Http/Middleware/CsrfProtectionMiddleware.php:
```diff
@@ -0,0 +1,184 @@
+        }
+
+        $method = $request->getMethod();
+        if ($method === 'requested') {
```
I don't think this works.

markstory commented on src/Http/Middleware/CsrfProtectionMiddleware.php:
```diff
@@ -0,0 +1,184 @@
+     * @param \Cake\Http\ServerRequest $request The request object.
+     * @return void
+     */
+    protected function _validateAndUnsetTokenField(ServerRequest &$request)
```
Use a return value and not pass by reference​.

markstory commented on src/Http/Middleware/CsrfProtectionMiddleware.php:
```diff
@@ -0,0 +1,184 @@
+            'secure' => $this->_config['secure'],
+            'httpOnly' => $this->_config['httpOnly'],
+        ]);
+    }
```
You're missing a return here.

burzum replied:
No, because `_setCookie(ServerRequest &$request, Response &$response)`. I've tried to return an array with both objects before but this leads into failing tests because the reference gets lost somewhere. Feel free to try it and remove the `&` and return this array:

```php
        return [
            'request' => $request,
            'response' => $response
        ];
```

burzum replied:
@markstory any feedback on this?

markstory replied:
All good now. :+1:

robertpustulka commented on src/Http/Middleware/CsrfProtectionMiddleware.php:
```diff
@@ -0,0 +1,184 @@
+
+        $params = $request->getAttribute('params');
+        $params['_csrfToken'] = $value;
+        $request = $request->withAttribute('params', $params);
```
This part does not "set cookie" but modifies the request, so it probably should be in another method.

inoas commented on src/Http/Middleware/CsrfProtectionMiddleware.php:
```diff
@@ -0,0 +1,184 @@
+/**
+ * Provides CSRF protection & validation.
+ *
+ * This component adds a CSRF token to a cookie. The cookie value is compared to
```
IMHO: `class CsrfProtectionCookieMiddleware` or `CsrfCookieMiddleware` or then?

inoas commented on src/Http/Middleware/CsrfProtectionMiddleware.php:
```diff
@@ -0,0 +1,184 @@
+     * @param callable $next Callback to invoke the next middleware.
+     * @return \Psr\Http\Message\ResponseInterface A response
+     */
+    public function __invoke(ServerRequestInterface &$request, ResponseInterface $response, $next)
```
Is pass by reference on purpose?

burzum replied:
Try to make asserts on the request object without it. No idea how else I could test it.

robertpustulka replied:
Expect method call on a mock?
```php
$request = $this->createMock(RequestInterface::class);
$request->expect('metod')->with($args)->willReturn($request);
```
?

burzum replied:
Got it to work by checking the asserts in the closure thanks to a friendly hint from @markstory

inoas commented on src/Http/Middleware/CsrfProtectionMiddleware.php:
```diff
@@ -0,0 +1,184 @@
+     */
+    protected function _validateAndUnsetTokenField(ServerRequest &$request)
+    {
+        if (in_array($request->getMethod(), ['PUT', 'POST', 'DELETE', 'PATCH']) || $request->getData()) {
```
IMHO always use `in_array($x, $y, true)` (as we always use `===`) unless there is a reason not to.

robertpustulka replied:
There are no falsey ot truey values in an array so it shouldn't be a problem?

inoas replied:
I am not saying it matters in this case. In many cases using empty()/isset() instead of === null or array_key_exists do not matter.

I am just saying this should be the default. We should always be strict with types UNLESS we have to be weak with it for whatever reason.

markstory commented on src/Http/Middleware/CsrfProtectionMiddleware.php:
```diff
@@ -0,0 +1,184 @@
+     * @param \Cake\Http\Response $response The response object.
+     * @return void
+     */
+    protected function _setCookie(ServerRequest &$request, Response &$response)
```
Another place we should avoid pass by ref.

markstory commented on tests/TestCase/Http/Middleware/CsrfProtectionMiddlewareTest.php:
```diff
@@ -0,0 +1,177 @@
+    public function tearDown()
+    {
+        parent::tearDown();
+    }
```
Do you need these no-op methods?

burzum committed:
Adding tests for the CsrfProtectionMiddleware

burzum committed:
Adding more tests for the CsrfProtectionMiddleware

burzum committed:
Removing method stubs from the CsrfProtectionMiddlewareTest

ADmad committed:
Update docblocks.

ADmad committed:
Don't pass by reference.

ADmad committed:
Fix CS error.

ADmad committed:
Remove cheeck for "requestAction()" requests.

We don't cater to deprecated features in new middleware. Thoses still
relying on requestAct() can keep using the CsrfComponent.

ADmad commented:
I have pushed commits to
- Update docblocks.
- Remove pass by reference in `_setToken()` method.
- Removed check/support for "requestAction()` requests.

markstory committed:
Stop using pass by ref and use smaller methods.

Pass by reference makes code hard to follow and enables multiple return
values which is counter intuitive.

markstory committed:
Stick to the interface methods.

Don't use methods outside of the typehinted interfaces.

markstory committed:
Fix phpcs error.

burzum committed:
Changing doc block of CsrfProtectionMiddleware::__invoke()

burzum commented:
@markstory @ADmad @dereuromark @robertpustulka so what is left to get this merged?

ADmad commented on src/Http/Middleware/CsrfProtectionMiddleware.php:
```diff
@@ -0,0 +1,197 @@
+     * @param \Cake\Http\Response $response The response to augment
+     * @return \Cake\Http\Response Modified response
+     */
+    protected function _addTokenCookie($token, $request, $response)
```
$request, $response params should perhaps be typehinted as done for other methods?

burzum committed:
Adding typehints to CsrfProtectionMiddleware::_addTokenCookie()

robertpustulka commented on src/Http/Middleware/CsrfProtectionMiddleware.php:
```diff
@@ -159,7 +159,7 @@ protected function _addTokenToRequest($token, ServerRequestInterface $request)
      * @param \Cake\Http\Response $response The response to augment
```
These should be changed as well.

burzum committed:
Updating a doc block in CsrfProtectionMiddleware

markstory reviewed (approved):
Looks good to me.
